### PR TITLE
Add memcached support

### PR DESCRIPTION
**Pull request by rkoster**
Added memcached support to auto-configure and cf-runtime gems.
Uses the [dalli](https://github.com/mperham/dalli) gem because it supports SASL and is [recommended](https://devcenter.heroku.com/articles/memcache#gem-installation) by heroku.

Bumped the cfoundry gem to a more recent version.

Added rake task for using cfruntime from source in cfruntime-tests. This to test unreleased changes.

### CODE REVIEW

File: auto-reconfiguration/Gemfile [hunk 1]
```diff
@@ -1,3 +1,5 @@
 source "http://rubygems.org"
 
 gemspec
+
+gemspec path: '../cfruntime'
```
Comment (drnic) [inline auto-reconfiguration/Gemfile]: This looks nice for your local dev; but probably won't work for anyone else?

Perhaps an "if" statement around it like https://github.com/cloudfoundry-community/bosh-bootstrap/blob/master/Gemfile#L5-L7?

File: auto-reconfiguration/Gemfile.lock [hunk 2]
```diff
@@ -2,11 +2,12 @@ PATH
   remote: .
   specs:
     cf-autoconfig (0.2.0)
+      cf-runtime (= 0.2.0)
 
 PATH
-  remote: ../cfruntime/
+  remote: /Users/rubenkoster/Projects/vcap-ruby/cfruntime
```
Comment (drnic) [inline auto-reconfiguration/Gemfile.lock]: Probably not the path where anyone else (e.g. travis) will find cfruntime :D

File: auto-reconfiguration/Gemfile (hunk 1 above)
Comment (rkoster) [inline auto-reconfiguration/Gemfile]: This should work because both gems are in the same repository. But It gives problems in the lockfile it seems.

Comment (drnic) [inline auto-reconfiguration/Gemfile]: Ahh its a project with multiple gems inside it. Got it.

File: auto-reconfiguration/Gemfile.lock (hunk 2 above)
Comment (drnic) [inline auto-reconfiguration/Gemfile.lock]: Given the above, I guess manually re-edit this line back to what it was.

OR, perhaps take the route that bosh took and have a single Gemfile for the whole project?

File: auto-reconfiguration/Gemfile [hunk 3]
```diff
@@ -1,3 +1,5 @@
 source "http://rubygems.org"
 
 gemspec
+
+gem 'cf-runtime', path: '../cfruntime'
```
Comment (mariash) [inline auto-reconfiguration/Gemfile:5]: Cf-Runtime is published gem and is required in gemspec already. https://github.com/cloudfoundry/vcap-ruby/blob/master/auto-reconfiguration/cf-autoconfig.gemspec#L18

**Comment by rkoster** [comment]
I know but because addeding memcached support required both gems to change I wantend to be able to test those changes before submitting a pull request. In general I don't think you should have to publish your gem before you can run specs.

**Comment by jbayer** [comment]
Ruben, 

Theoretically, I'm in support of memcache service auto-configuration. However, there are 2 problems.

1) The CF eng team dooesn't have a memcache server to test with
2) This PR will bring in some LGPL v3 code via kgio http://bogomips.org/kgio/LICENSE.html

Not having a test env could be ok. However, earlier this week, we decided to not use Sidekiq for background jobs, at least partially due to licensing considerations for the wider community including IBM and others not comfortable with LGPL v3.
https://groups.google.com/a/cloudfoundry.org/d/msg/vcap-dev/5GRSCpFjsYs/0NMSH4pJhc0J

Let me know if you have any suggestions.

**Comment by rkoster** [comment]
@jbayer I'm unable to find a reference to the kgio gem. The only gem needed for memcached support is [dalli](http://rubygems.org/gems/dalli) which is licensed under MIT. This is the same gem used by the [memcached service](https://github.com/cloudfoundry/cf-services-contrib-release/blob/master/src/services/ng/memcached/Gemfile#L11).

**Comment by RobDay-Reynolds** [comment]
dalli's Gemfile includes kgio, and such nested dependencies may cause issues.  cf-services-contrib-release is a community repo, and not a part of our core repository, so licensing considerations are different there.  We are reviewing the license restrictions now, and will update as we hear more. Overall, we reviewed pull request and it looks good.

@monkeyherder & @mariash

**Comment by rkoster** [comment]
The kgio gem is indeed listed in the [Gemfile](https://github.com/mperham/dalli/blob/master/Gemfile#L6) this file however is not used for declaring runtime dependencies. Runtime dependencies sould be declared in de [.gemspec](https://github.com/mperham/dalli/blob/master/dalli.gemspec) file. In the case of the dalli gem there are however none. So only code, licensed under MIT will be installed as a result of the use of this gem.

**Comment by jbayer** [comment]
Ok, I'm ignorant in this area, and we have to get some legal advice from our more knowledgeable team.

**Comment by rkoster** [comment]
@jbayer any updates from the legal team? Because it would be nice if we know if this can be merged before we start adding other services.

**Comment by jbayer** [comment]
@rkoster sorry i've been on vacation and am still catching up. one of our lawyers sent us some feedback and i need to clarify some things with her.

**Comment by ryantang** [comment]
An update on this PR - we are looking for an internal engineering team to take ownership of this repo since it is an orphan currently. The difficulty is around validation and support- we don't have a memcache setup to test this with so our team would have to take on this additional work. If no one steps up internally, we may ask someone in the community to take it over. 

@jbayer said that there are no legal issues around pulling this code in.

@jfoley / @ryantang

**Comment by xoebus** [comment]
@shalako — we understand you're looking into this repository.  Can you comment on this PR?

@ryantang / @xoebus

**Comment by shalako** [comment]
No update since Ryan's latest post, I'm running with it for now.

**Comment by andreasmaier** [comment]
@mkocher Have you had a chance to look at this?

**Comment by shalako** [comment]
An investigation of vcap_ruby is in @dsboulder's hands at the moment.

**Comment by jbayer** [comment]
@rkoster @drnic we've decided based on the advice of @dsboulder and my agreement that we'd like to discontinue supporting this repository. our opinionated position is that we should have a simple library that only parses the VCAP_SERVICES values easily and possibly offers some helper functions like https://github.com/cloudfoundry/vcap-ruby/blob/master/cfruntime/lib/cf-runtime/properties.rb#L10

do you think the cloudfoundry-community org would like to support this codebase/repository?

**Comment by rkoster** [comment]
I think it is important this repo gets maintained at least until all the service client gems support `VCAP_SERVICES`. For example to simple configure `dalli` without using this gem you would need to create a dynamic config file similar to the [dynamic database.yml](https://github.com/cloudfoundry/heroku-buildpack-ruby/blob/master/lib/language_pack/ruby.rb#L504-L566)
And then do a [dirty hack](https://support.cloud.engineyard.com/entries/22375358-Using-Memcached-on-Engine-Yard-Cloud) to get the file loaded.

**Comment by jbayer** [comment]
@drnic ping. are you open to taking this repository in cloudfoundry-community?

**Comment by drnic** [comment]
Yes

**Comment by drnic** [comment]
James, you're now an Owner of cf-community org. You can transfer the repo
if you'd like.

Nic

On Mon, Sep 16, 2013 at 7:48 PM, Dr Nic Williams drnicwilliams@gmail.comwrote:

> Yes

## 

Dr Nic Williams
Stark & Wayne LLC - consultancy for Cloud Foundry users
http://drnicwilliams.com
http://starkandwayne.com
cell +1 (415) 860-2185
twitter @drnic

**Comment by elliterate** [comment]
We've deprecated this repository in favor of [cloudfoundry/cf-app-utils-ruby](https://github.com/cloudfoundry/cf-app-utils-ruby).

The old code has been moved to [cloudfoundry-community/vcap-ruby](https://github.com/cloudfoundry-community/vcap-ruby). If desired, please open a new pull request for this change over there.

**Pull request closed without merging**